### PR TITLE
Enhancement/remove onlinelibraryfragment

### LINKED ISSUE #4612: Remove `OnlineLibraryFragment` from project, and use pure composable instead.
Parent Issue https://github.com/kiwix/kiwix-android/issues/4572

### PR DESCRIPTION
**Pull request by yush-1018**
Fixes #4612

### Architecture & Refactoring
* Extracted business logic from OnlineLibraryViewModel into dedicated classes:
  * `ObserveOnlineLibrary`
  * `ObserveOnlineLibraryItems`
  * `ObserveNetworkState`
  * `ResolveBookClickAction`
  * `ResolveRefreshLibraryAction`
* Introduced OnlineLibraryRepository and supporting mappers for better separation of concerns.
* Simplified ViewModel by using a single UiState and reactive flows.
* Moved OPDS-related logic to the app module to improve modularity.
* Removed `OnlineLibraryFragment` from the project and migrated all the functionalities into pure Compose.
* Using `Accompanist` for permission handling.
* Refactored `AvailableSpaceCalculator` to return a typed result instead of using success and failure lambdas, improving readability and testability.

### Performance Improvements
* Injected Downloader using Provider for lazy initialization.
  * Prevents unnecessary creation during ViewModel init
  * Avoids Fetch’s SQLite queries running on the main thread
  * Reduces initial loading overhead

### Testing Improvements
* Added comprehensive unit tests for:
  * `OnlineLibraryViewModel`
  * `ObserveOnlineLibrary`, `ObserveOnlineLibraryItems`, `ObserveNetworkState`
  * `ResolveBookClickAction`, `ResolveRefreshLibraryAction`
  * `OnlineLibraryRepository`
* Added edge case coverage (pagination, permissions, network states, initial loading behavior, permission flows, etc).
* Added tests for lazy Downloader initialization via Provider.
* Introduced Compose UI tests for OnlineLibraryScreen.
* Refactored and stabilized existing flaky tests.

### UX Improvements
* Improved loading and parsing messages.
* Better handling of empty states and network scenarios.
* Cleaner and more predictable search behavior.

### CODE REVIEW

**Comment by yush-1018** [comment]
Hi, @MohitMaliDeveloper
CI failures are due to flaky emulator tests (storage full on API 25, compose timeout on API 34). Same tests fail on other PRs too. All 17 other checks pass. Could you please re-run the failed jobs? Thanks!

File: app/src/main/java/org/kiwix/kiwixmobile/di/components/KiwixActivityComponent.kt [hunk 1]
```diff
@@ -49,9 +49,15 @@ interface KiwixActivityComponent : CoreActivityComponent {
   fun inject(localFileTransferFragment: LocalFileTransferFragment)
   fun inject(zimHostFragment: ZimHostFragment)
   fun inject(kiwixMainActivity: KiwixMainActivity)
-  fun inject(onlineLibraryFragment: OnlineLibraryFragment)
+
   fun inject(onlineCategoryDialog: OnlineCategoryDialog)
 
+  fun connectivityManager(): android.net.ConnectivityManager
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/di/components/KiwixActivityComponent.kt]: These can be directly injected into the viewModel. We don't need to define them.

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineLibraryRoute.kt [hunk 2]
```diff
@@ -0,0 +1,873 @@
...
+@Composable
+fun OnlineLibraryRoute(
+  viewModelFactory: ViewModelProvider.Factory,
+  alertDialogShower: AlertDialogShower,
+  navController: NavHostController
+) {
+  val context = LocalContext.current
+  val activity = context as KiwixMainActivity
+  val component = activity.cachedComponent
+  val downloader = component.downloader()
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineLibraryRoute.kt]: These all should come from the viewModel.

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineLibraryRoute.kt [hunk 3]
```diff
@@ -0,0 +1,873 @@
...
+private fun ObserveLibraryState(
+  routeState: OnlineLibraryRouteState,
+  lifecycleOwner: LifecycleOwner,
+  onRefreshingChanged: (Boolean) -> Unit,
+  onScanningProgressChanged: (Pair<Boolean, String>) -> Unit,
+  onNoContentChanged: (Pair<String, Boolean>) -> Unit
+) {
+  DisposableEffect(lifecycleOwner) {
+    val observer = Observer<Boolean> { shouldShow ->
+      if (shouldShow == true && !NetworkUtils.isWiFi(routeState.context)) {
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineLibraryRoute.kt]: `shouldShow` is not a nullable type, so comparing is redundant here. We can remove the `== true`

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineLibraryRoute.kt [hunk 4]
```diff
@@ -0,0 +1,873 @@
...
+    routeState.zimManageViewModel.shouldShowWifiOnlyDialog.observe(lifecycleOwner, observer)
+    onDispose {
+      routeState.zimManageViewModel.shouldShowWifiOnlyDialog.removeObserver(observer)
+    }
+  }
+
+  ObserveNetworkState(routeState, lifecycleOwner, onNoContentChanged, onScanningProgressChanged)
+
+  DisposableEffect(lifecycleOwner) {
+    val observer = Observer<Boolean> { refreshing -> onRefreshingChanged(refreshing == true) }
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineLibraryRoute.kt]: Here too, we should remove the `== true`.

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineLibraryRoute.kt [hunk 5]
```diff
@@ -0,0 +1,873 @@
...
+    val decorView = activity.window.decorView
+    val insets = androidx.core.view.ViewCompat.getRootWindowInsets(decorView)
+    val isKeyboardVisible =
+      insets?.isVisible(androidx.core.view.WindowInsetsCompat.Type.ime()) == true
+    if (isKeyboardVisible || isSearchActive) {
+      activity.currentFocus?.let { view ->
+        val imm = activity.getSystemService(
+          android.content.Context.INPUT_METHOD_SERVICE
+        ) as android.view.inputmethod.InputMethodManager
+        imm.hideSoftInputFromWindow(view.windowToken, 0)
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineLibraryRoute.kt]: There is an extension function for closing the keyboard. We should use that.

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineLibraryRoute.kt [hunk 6]
```diff
@@ -0,0 +1,873 @@
...
+            handleNetworkConnected(routeState, onNoContentChanged, onScanningProgressChanged)
+          }
+          NetworkState.NOT_CONNECTED -> {
+            showNoInternetConnectionError(routeState)
+          }
+          else -> {}
+        }
+      }
+    }
+    routeState.zimManageViewModel.networkStates.observe(lifecycleOwner, observer)
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineLibraryRoute.kt]: Now, we are migrating to pure composable, so we should use the compose way to observe the data. So we should use `MutableStateFlow`, `MutableSharedFlow` instead of `MutableLiveData`. They are already lifecycle aware.

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineLibraryRoute.kt [hunk 7]
```diff
@@ -0,0 +1,873 @@
...
+    onDispose {
+      routeState.zimManageViewModel.libraryListIsRefreshing.removeObserver(observer)
+    }
+  }
+
+  DisposableEffect(lifecycleOwner) {
+    val observer = Observer<String> { status ->
+      onScanningProgressChanged(true to status)
+    }
+    routeState.zimManageViewModel.downloadProgress.observe(lifecycleOwner, observer)
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineLibraryRoute.kt]: Same for this and all other data.

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineLibraryRoute.kt [hunk 8]
```diff
@@ -0,0 +1,873 @@
...
+private fun rememberStoragePermissionLauncher(
+  routeState: OnlineLibraryRouteState,
+  alertDialogShower: AlertDialogShower,
+  activity: KiwixMainActivity
+): ActivityResultLauncher<String> =
+  rememberLauncherForActivityResult(ActivityResultContracts.RequestPermission()) { isGranted ->
+    if (isGranted) {
+      clickOnBookItem(routeState)
+    } else {
+      val showRationale = ActivityCompat.shouldShowRequestPermissionRationale(
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineLibraryRoute.kt]: We have the `KiwixPermissionChecker` and we are managing all the permission checking there. You should use that in the viewModel for checking the permission instead of directly placing checks in the UI.

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineLibraryRoute.kt [hunk 9]
```diff
@@ -0,0 +1,873 @@
...
+private fun downloadFile(routeState: OnlineLibraryRouteState) {
+  routeState.downloadBookItem?.book?.let {
+    routeState.downloader.download(it)
+    routeState.setDownloadBookItem(null)
+  }
+}
+
+private fun clickOnBookItem(routeState: OnlineLibraryRouteState) {
+  routeState.scope.launch {
+    if (!routeState.activity.isManageExternalStoragePermissionGranted(routeState.kiwixDataStore)) {
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineLibraryRoute.kt]: Move this permission check inside the `KiwixPermissionChecker` so that we can use it in all over the application.

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineLibraryRoute.kt [hunk 10]
```diff
@@ -0,0 +1,873 @@
...
+    }
+    onLoadingMoreChanged(loadingMoreItem)
+  }
+}
+
+private fun getOnlineLibraryRequest(
+  kiwixDataStore: org.kiwix.kiwixmobile.core.utils.datastore.KiwixDataStore
+): OnlineLibraryRequest = OnlineLibraryRequest(
+  null,
+  runBlocking {
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineLibraryRoute.kt]: Now, we are properly migrating into compose. It is time to remove the `runBlocking`. Make this method suspend since it is called from `LaunchedEffect`.

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineLibraryRoute.kt [hunk 11]
```diff
@@ -0,0 +1,873 @@
...
+    routeState.setDownloadBookItem(item)
+    if (routeState.activity.hasNotificationPermission(routeState.kiwixDataStore)) {
+      when {
+        !NetworkUtils.isNetworkAvailable(routeState.activity) -> {
+          noInternetSnackbar(routeState)
+          return@launch
+        }
+        routeState.kiwixDataStore.wifiOnly.first() &&
+          !NetworkUtils.isWiFi(routeState.context) -> {
+          routeState.alertDialogShower.show(KiwixDialog.YesNoDialog.WifiOnly, {
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineLibraryRoute.kt]: This method is heavily doing things. We should split this. Dialogs should show via effects, not directly. See the `SettingsScreen` implementation.

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineLibraryRoute.kt [hunk 12]
```diff
@@ -0,0 +1,873 @@
...
+private fun checkExternalStorageWritePermissionAndProceed(
+  routeState: OnlineLibraryRouteState,
+  storagePermissionLauncher: ActivityResultLauncher<String>
+) {
+  routeState.scope.launch {
+    if (!routeState.kiwixDataStore.isPlayStoreBuildWithAndroid11OrAbove()) {
+      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+        clickOnBookItem(routeState)
+      } else {
+        storagePermissionLauncher.launch(Manifest.permission.WRITE_EXTERNAL_STORAGE)
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineLibraryRoute.kt]: Why are we directly launching the permission launcher without checking whether the permission is already granted or not? On Android 13 and below, it will launch this this everytime when a book item is clicked.

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineLibraryRoute.kt [hunk 13]
```diff
@@ -0,0 +1,873 @@
...
+@Suppress("NestedBlockDepth")
+private fun onBookItemClick(
+  routeState: OnlineLibraryRouteState,
+  item: LibraryListItem.BookItem,
+  storagePermissionLauncher: ActivityResultLauncher<String>,
+  notificationPermissionLauncher: ActivityResultLauncher<String>
+) {
+  routeState.scope.launch {
+    routeState.setDownloadBookItem(item)
+    if (routeState.activity.hasNotificationPermission(routeState.kiwixDataStore)) {
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineLibraryRoute.kt]: This permission check should be inside the KiwixPermissionChecker.

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineLibraryRoute.kt [hunk 14]
```diff
@@ -0,0 +1,873 @@
...
+        ) as android.view.inputmethod.InputMethodManager
+        imm.hideSoftInputFromWindow(view.windowToken, 0)
+      }
+    },
+    isEnabled = true,
+    testingTag = LANGUAGE_MENU_ICON_TESTING_TAG
+  )
+)
+
+private fun handleLoadMore(zimManageViewModel: ZimManageViewModel, count: Int) {
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineLibraryRoute.kt]: This should be handled in the viewModel, not on the UI.

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineLibraryRoute.kt [hunk 15]
```diff
@@ -0,0 +1,873 @@
...
+private fun showManageExternalStoragePermissionDialog(routeState: OnlineLibraryRouteState) {
+  if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
+    routeState.alertDialogShower.show(
+      KiwixDialog.ManageExternalFilesPermissionDialog,
+      { routeState.activity.let(FragmentActivity::navigateToSettings) }
+    )
+  }
+}
+
+private fun showNoSpaceSnackbar(routeState: OnlineLibraryRouteState, availableSpace: String) {
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineLibraryRoute.kt]: Instead of directly showing the snackbar and dialogs. We should emit `events` and listen to them on the UI.

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineLibraryRoute.kt [hunk 16]
```diff
@@ -0,0 +1,873 @@
...
+  var isSearchActive by remember { mutableStateOf(false) }
+  var searchText by remember { mutableStateOf("") }
+  var isLoadingMoreItem by remember { mutableStateOf(false) }
+  var scanningProgressItem by remember { mutableStateOf(Pair(false, "")) }
+  var noContentViewItem by remember { mutableStateOf(Pair("", false)) }
+  var isRefreshing by remember { mutableStateOf(false) }
+  var onlineLibraryList by remember { mutableStateOf<List<LibraryListItem>?>(null) }
+  var downloadBookItem: LibraryListItem.BookItem? by remember { mutableStateOf(null) }
+
+  val routeState = OnlineLibraryRouteState(
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineLibraryRoute.kt]: After moving to the viewModel approach i guess we can get rid of this.

**Review by MohitMaliFtechiz — CHANGES_REQUESTED** [review]
@yush-1018 Thank you for making the changes.


The download is not working on the first click. I have to click twice to start the download. It is not a server problem. I am not seeing the logs for requesting the download on the first click.

https://github.com/user-attachments/assets/16b2aaa9-8b7d-4cf4-bc09-c4bd02c04a7d


Also, now we are splitting the code into seperate viewModels for localLibrary and onlineLibrary. So we should use that approach here also.

**Comment by yush-1018** [comment]
Hi, @MohitMaliFtechiz 

I've addressed the comments, including replacing LiveData with StateFlow, using suspend functions instead of runBlocking, and improving permission handling. 

Please review for any changes !

Thanku 😊

**Comment by yush-1018** [comment]
Hi @MohitMaliFtechiz ,
Can u please confirm the checks.

Thanks !

**Comment by MohitMaliFtechiz** [comment]
@yush-1018 Thank you for making the changes. But the CI is failing. Could you please fix the test failure? Something is happening, because CI is failing on `DownloadTest`. Once the CI is fixed i will review the PR. Because it takes more time than usual to review PR like this.

**Comment by atharvyadav22** [comment]
@yush-1018 
Should i help you with this migration i see many issues and migration is important so we need to handle it.

**Comment by atharvyadav22** [comment]
@MohitMaliFtechiz can i take this if @yush-1018 is okay with it?

**Comment by yush-1018** [comment]
I still working on this.

Thanks !

**Comment by MohitMaliFtechiz** [comment]
For asking permission, use the `accompanist` library. See #4819.

**Comment by MohitMaliFtechiz** [comment]
@harshsomankar123-tech Are you still working on this? Or should I take this one?

**Comment by harshsomankar123-tech** [comment]
@MohitMaliFtechiz You can take this, I’ve already done most of the work.Only the CI failure remains.
Thanks!

**Comment by kelson42** [comment]
@MohitMaliDeveloper What is the status here? It seems thisnPR expects a new review pass!?

**Comment by MohitMaliFtechiz** [comment]
> @MohitMaliDeveloper What is the status here? It seems thisnPR expects a new review pass!?

@kelson42 I am adding the new test cases and refactoring the existing test cases(there are many architecture changes that have been made in this PR, which need to be tested properly. Also i will add the pure compose UI test cases as well. Once done, I will mark this PR for review.

**Comment by MohitMaliFtechiz** [comment]
@harshsomankar123-tech, @atharvyadav22 The code refactore + test cases are done. This is a very important part of our application. I request you both to test and review the PR, and give your feedback :)

**Comment by atharvyadav22** [comment]
@MohitMaliFtechiz 

### Issue 1
 Few Of the books don't download 

https://github.com/user-attachments/assets/023829a7-192f-41b4-b570-6eb7a96dc948

https://github.com/user-attachments/assets/cafd09c8-cfb7-43da-b17d-56c5db8f01f9

Logs-

image


### Issue 2
Category Don't show loading 

https://github.com/user-attachments/assets/bff2f813-9fdc-4bfa-a3a4-df7f0b31db6a

### Issue 3
When you select an particular category (other than All category) and then try to chsnge language there is no updation also there is no loading indicator visible

https://github.com/user-attachments/assets/84249672-8b81-44ff-81cf-1467f8630512

Note: - Its i UI updation issue confirmed via logs

image

**Comment by MohitMaliFtechiz** [comment]
@atharvyadav22 The first issue: it clearly shows the URL is not found on the server. I will open an issue for this.

Second and third issues: I am fixing this.

Apart from this, have you found any other issue?

**Comment by atharvyadav22** [comment]
> @atharvyadav22 The first issue: it clearly shows the URL is not found on the server. I will open an issue for this.
> 
> Second and third issues: I am fixing this.
> 
> Apart from this, have you found any other issue?

For behavior no such issues found once i go through the code ill check that functionality

**Comment by MohitMaliFtechiz** [comment]
> For behavior no such issues found once i go through the code ill check that functionality

Okay. 

* Test it on the emulator as well, where 2 or more storage are available.
* Permission-related things on Android 13 and above, and above as well.
* When not enough storage is available on the selected storage, it should show the storageSelectionDialog.

**Comment by MohitMaliFtechiz** [comment]
@atharvyadav22 For the first issue, I have opened an issue https://github.com/kiwix/kiwix-android/issues/4858. For the rest of the issues, I have placed a fix.

**Comment by atharvyadav22** [comment]
### Issue 4
Memory Leaks

Screenshot_2026-05-07-21-22-23-45


https://github.com/user-attachments/assets/116a88b4-0eac-4054-93db-61e25cdd9b9b

### Issue 5
No Items Available text appears during Loading Indicator mainly State Issue

https://github.com/user-attachments/assets/21f12ccf-39be-4d87-87db-759988e94083

**Comment by MohitMaliFtechiz** [comment]
The memory leak is caused by our service running while downloading the ZIM files. We will handle this separately. For the "No items available" I have placed a fix.

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/viewmodel/OnlineLibraryViewModel.kt [hunk 17]
```diff
@@ -0,0 +1,778 @@
...
+
+  private fun updateOnlineLibraryFilters(newRequest: OnlineLibraryRequest) {
+    val updatedRequest = currentRequest.copy(
+      query = newRequest.query ?: currentRequest.query,
+      category = newRequest.category ?: currentRequest.category,
+      lang = newRequest.lang ?: currentRequest.lang,
+      page = newRequest.page,
+      isLoadMoreItem = newRequest.isLoadMoreItem
+    )
+    if (updatedRequest == currentRequest) return
```
[the remark this answers is not in the document]
Comment (harshsomankar123-tech) [reply app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/viewmodel/OnlineLibraryViewModel.kt]: I Think this is still valid. The same equality guard also breaks explicit refresh: `refreshScreen(true)` calls `updateOnlineLibraryFilters(getOnlineLibraryRequest())`, so unchanged filters return here without emitting to `onlineLibraryRequest`, while the UI can still show the loading state. Please fix this with a force-refresh path / same-request emission and add a regression test.

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/viewmodel/OnlineLibraryViewModel.kt [hunk 18]
```diff
@@ -0,0 +1,799 @@
...
+import org.kiwix.libkiwix.Book
+import javax.inject.Inject
+
+/**
+ * ViewModel for the OnlineLibraryRoute composable.
+ * Holds dependencies and business logic, emitting UI events for the composable to handle.
+ */
+@Suppress("LongParameterList")
+class OnlineLibraryViewModel @Inject constructor(
+  val downloader: Downloader,
```
Comment (harshsomankar123-tech) [inline app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/viewmodel/OnlineLibraryViewModel.kt]: Injecting Downloader here eagerly initializes the whole dependency chain (Downloader -> DownloadRequester -> Fetch) during ViewModel creation. Based on the StrictMode trace, this eventually calls Fetch.getInstance() and opens Fetch’s SQLite DB on the main thread while entering the Downloads screen.

Could we make this dependency lazy/provider-backed instead, so Fetch is only initialized when a download action is actually triggered? Also, the first Fetch initialization should happen from an IO context to avoid main-thread disk access.WDYT?

**Review by harshsomankar123-tech — CHANGES_REQUESTED** [review]
Thanks @MohitMaliFtechiz. I checked the latest changes and reviewed the Logcat output while opening and refreshing the online library screen. I found two issues that need to be addressed before this can merge.PTAL

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/viewmodel/OnlineLibraryViewModel.kt (hunk 18 above)
Comment (MohitMaliFtechiz) [reply app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/online/viewmodel/OnlineLibraryViewModel.kt]: Thank you for the idea and review. I have implemented both requests in the code.

> I Think this is still valid. The same equality guard also breaks explicit refresh: refreshScreen(true) calls updateOnlineLibraryFilters(getOnlineLibraryRequest()), so unchanged filters return here without emitting to onlineLibraryRequest, while the UI can still show the loading state. Please fix this with a force-refresh path / same-request emission and add a regression test.

The best is to remove this condition. Because updating the library only triggers once. So there are now no duplicate requests. For the reason we have implemented this(initial library load), which is already fixed, so we have removed this.

**Comment by MohitMaliFtechiz** [comment]
@harshsomankar123-tech Does the last fix for your request look good to you? Apart from that, have you faced any other issues?

**Review by harshsomankar123-tech — DISMISSED** [review]
@MohitMaliFtechiz Thank U so much for working on this! Now this looks good to me. You’ve safely migrated the Fragment to Compose and also added tests covering all the logic.

**Comment by MohitMaliFtechiz** [comment]
@kelson42 Now, this PR is ready for merge.

**Comment by atharvyadav22** [comment]
@MohitMaliFtechiz 
Should we have a look at this too since we are covering migration #4698

**Comment by MohitMaliFtechiz** [comment]
> Since ZIM don't change continuously it should not make API call everytime as it costs additional data.

@atharvyadav22 The first point of the issue is already fixed in this PR. The second point is changing the UI, which should not be inside this PR.

**Comment by kelson42** [comment]
> > Since ZIM don't change continuously it should not make API call everytime as it costs additional data.
> 
> @atharvyadav22 The first point of the issue is already fixed in this PR. The second point is changing the UI, which should not be inside this PR.

then please open/link issue so we don't loose track?

**Comment by harshsomankar123-tech** [comment]
@kelson42 @MohitMaliFtechiz @atharvyadav22 I think we can merge this one. The issue still needs some discussion, but since this is a big PR, it’s blocking other tickets.WDUT??
Thanks!

**Comment by MohitMaliFtechiz** [comment]
@harshsomankar123-tech Agree with you. https://github.com/kiwix/kiwix-android/issues/4698 needs discussion since it is a UI change(and the showing progress of fetching the online library will be gone). This PR already has many changes, and the purpose of this PR is to migrate to compose(not changing the UI and behaviour).

> it’s blocking other tickets.WDUT

Yes, other PRs like https://github.com/kiwix/kiwix-android/pull/4705, https://github.com/kiwix/kiwix-android/pull/4746, and https://github.com/kiwix/kiwix-android/pull/4782 depend on this one.
